examples/ecs/observers.rs

```diff
@@ -0,0 +1,198 @@
+//! This examples illustrates several ways you can employ observers
+
+use bevy::{
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_systems(Startup, setup)
+        .add_systems(Update, (draw_shapes, handle_click))
+        .run();
+}
+
+#[derive(Component)]
+struct Mine {
+    pos: Vec2,
+    size: f32,
+}
+
+#[derive(Component)]
+struct TriggerMines {
+    pos: Vec2,
+    radius: f32,
+}
+
+#[derive(Component)]
+struct Explode;
+
+fn setup(world: &mut World) {
+    world.spawn(Camera2dBundle::default());
+
+    let font = world
+        .resource::<AssetServer>()
+        .load("fonts/FiraMono-Medium.ttf");
+    world.spawn(TextBundle::from_section(
+        "Click on a \"Mine\" to trigger it.\n\
+            When it explodes it will trigger all overlapping mines.",
+        TextStyle {
+            font,
+            font_size: 24.,
+            color: Color::WHITE,
+        },
+    ));
+
+    // Pre-register all our components, resource and event types.
+    world.init_resource::<SpatialIndex>();
+    world.init_component::<Mine>();
+    world.init_component::<TriggerMines>();
+    world.init_component::<Explode>();
+
+    // Observers are triggered when a certain event it fired, each event is represented by a component type.
+    // This observer runs whenever `TriggerMines` is fired, observers run systems which can be defined as closures.
+    world.observer(
+        |observer: Observer<TriggerMines>,
+         mines: Query<&Mine>,
+         index: Res<SpatialIndex>,
+         mut commands: Commands| {
+            // You can access the event data via the `Observer`
+            let trigger = observer.data();
+            // Access resources
+            for e in index.get_nearby(trigger.pos) {
+                // Run queries
+                let mine = mines.get(e).unwrap();
+                if mine.pos.distance(trigger.pos) < mine.size + trigger.radius {
+                    // And queue commands, including firing additional events
+                    // Here we fire the `Explode` event at entity `e`
+                    commands.event(Explode).entity(e).emit();
```

Comment:
Yeah we might be able to use `Drop` here, will give it a go in the next pass.

Reply:
This is problematic as the builder borrows world and if it implemented `Drop` would not drop the borrow until end of scope, a similar problem to what we faced with adding a `Drop` impl to `EntityWorldMut`.

Happy to revisit this if it becomes a pain point

Reply:
If adding a `Drop` impl is something we want, we should do so in the first implementation. It's very difficult to add later on, due to the reasons you mentioned. You can get around the scope issues by having a method to emit the event early and release the borrow

Reply:
Personally I prefer the explicitness of `#[must_use]` rather than implicitly omitting the emit call in most cases but sometimes requiring you to call it. I can change it though if `Drop` is preferred.

Reply:
Yeah, I like the explicit `must_use` better here. I think it's more clear, more flexible and causes fewer surprising borrow-checker issues.